### PR TITLE
feat: new module `avm/res/network/nat-gateway`

### PR DESCRIPTION
**Pull request by fabmas**
## Description

New network/nat-gateway module migrated from CARML to AVM

[avm.res.network.nat-gateways](https://github.com/fabmas/bicep-registry-modules/actions/workflows/avm.res.network.nat-gateway.yml)

### CODE REVIEW

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]
Requires update to codeowners file (& potential creation of teams)

File: avm/res/network/nat-gateway/tests/e2e/max/main.test.bicep [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    }
+    publicIPAddressObjects: [
+      {
+        name: '${namePrefix}${serviceShort}001-pip'
+        roleAssignments: [
+          {
+            roleDefinitionIdOrName: 'Reader'
+            principalId: nestedDependencies.outputs.managedIdentityPrincipalId
+            principalType: 'ServicePrincipal'
+          }
```
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/tests/e2e/max/main.test.bicep:84]:
```suggestion
      {
        roleDefinitionIdOrName: 'Owner'
        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
        principalType: 'ServicePrincipal'
      }
      {
        roleDefinitionIdOrName: 'b24988ac-6180-42a0-ab88-20f7382dd24c'
        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
        principalType: 'ServicePrincipal'
      }
      {
        roleDefinitionIdOrName: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')
        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
        principalType: 'ServicePrincipal'
      }
```

As reference for other users.

File: avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        roleDefinitionIdOrName: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')
+        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
+        principalType: 'ServicePrincipal'
+      }
+    ]
+    tags: {
+      'hidden-title': 'This is visible in the resource name'
+      Environment: 'Non-Prod'
+      Role: 'DeploymentValidation'
+    }
```
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep:117]:
```suggestion
```

Should be removed as unrelated to test scenario and covered by another test case. If also the other role assignment is removed it enables you to remove the dependency.

File: avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+param resourceGroupName string = 'dep-${namePrefix}-network.natgateway-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'nngcprx'
+
+@description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
+param enableTelemetry bool = true
```
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep:21]:
```suggestion
```

File: avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+// Test Execution //
+// ============== //
+
+@batchSize(1)
+module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem' ]: {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
+  params: {
+    location: location
+    enableTelemetry: enableTelemetry
```
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep:71]:
```suggestion
```

File: avm/res/network/nat-gateway/tests/e2e/waf-aligned/main.test.bicep [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    publicIPAddressObjects: [
+      {
+        name: '${namePrefix}${serviceShort}001-pip'
+        roleAssignments: [
+          {
+            roleDefinitionIdOrName: 'Reader'
+            principalId: nestedDependencies.outputs.managedIdentityPrincipalId
+            principalType: 'ServicePrincipal'
+          }
+        ]
```
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/tests/e2e/waf-aligned/main.test.bicep:85]:
```suggestion
```

Should be removed as it's not required to be WAF-aligned. You can also remove the dependency

File: avm/res/network/nat-gateway/tests/e2e/waf-aligned/main.test.bicep [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+// Test Execution //
+// ============== //
+
+@batchSize(1)
+module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem' ]: {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
+  params: {
+    location: location
+    enableTelemetry: enableTelemetry
```
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/tests/e2e/waf-aligned/main.test.bicep:70]:
```suggestion
```

File: avm/res/network/nat-gateway/tests/e2e/waf-aligned/main.test.bicep [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+param resourceGroupName string = 'dep-${namePrefix}-network.natgateway-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'nngwaf'
+
+@description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
+param enableTelemetry bool = true
```
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/tests/e2e/waf-aligned/main.test.bicep:21]:
```suggestion
```

File: avm/res/network/nat-gateway/tests/e2e/max/main.test.bicep [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+param resourceGroupName string = 'dep-${namePrefix}-network.natgateway-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'nngmax'
+
+@description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
+param enableTelemetry bool = true
```
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/tests/e2e/max/main.test.bicep:21]:
```suggestion
```

File: avm/res/network/nat-gateway/tests/e2e/max/main.test.bicep [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+// Test Execution //
+// ============== //
+
+@batchSize(1)
+module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem' ]: {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
+  params: {
+    location: location
+    enableTelemetry: enableTelemetry
```
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/tests/e2e/max/main.test.bicep:70]:
```suggestion
```

File: avm/res/network/nat-gateway/main.bicep [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+        telemetry: {
+          type: 'String'
+          value: 'For more information, see https://aka.ms/avm/TelemetryInfo'
+        }
+      }
+    }
+  }
+}
+
+module publicIPAddresses '../public-ip-address/main.bicep' = [for (publicIPAddressObject, index) in (publicIPAddressObjects ?? []): {
```
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/main.bicep:70]: Must reference the published module

File: avm/res/network/nat-gateway/main.bicep [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+@description('Optional. Array of role assignments to create.')
+param roleAssignments roleAssignmentType
+
+@description('Optional. Tags for the resource.')
+param tags object?
+
+@description('Optional. Enable/Disable usage telemetry for module.')
+param enableTelemetry bool = true
+
+var enableReferencedModulesTelemetry = false
```
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/main.bicep:41]:
```suggestion
```

File: .github/workflows/avm.res.network.nat-gateway.yml
Comment (eriqua) [inline .github/workflows/avm.res.network.nat-gateway.yml:1]: Please uncomment related line in CODEOWNERS file

File: .github/CODEOWNERS [hunk 12]
```diff
@@ -93,7 +93,7 @@
 #/avm/res/network/ip-group/ @Azure/avm-res-network-ipgroup-module-owners-bicep @Azure/avm-core-team-technical-bicep
 /avm/res/network/load-balancer/ @Azure/avm-res-network-loadbalancer-module-owners-bicep @Azure/avm-core-team-technical-bicep
 #/avm/res/network/local-network-gateway/ @Azure/avm-res-network-localnetworkgateway-module-owners-bicep @Azure/avm-core-team-technical-bicep
-#/avm/res/network/nat-gateway/ @Azure/avm-res-network-natgateway-module-owners-bicep @Azure/avm-core-team-technical-bicep
+/avm/res/network/nat-gateway/ @Azure/avm-res-network-natgateway-module-owners-bicep @Azure/avm-core-team-technical-bicep
```
Comment (eriqua) [inline .github/CODEOWNERS:96]: The `avm-res-network-natgateway-module-owners-bicep` team should be added as member of `avm-technical-reviewers-bicep`

Comment (eriqua) [reply .github/CODEOWNERS:96]: Solved

File: avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-diagnosticDependencies'
+  params: {
+    storageAccountName: 'dep${namePrefix}diasa${serviceShort}01'
+    logAnalyticsWorkspaceName: 'dep-${namePrefix}-law-${serviceShort}'
+    eventHubNamespaceEventHubName: 'dep-${namePrefix}-evh-${serviceShort}'
+    eventHubNamespaceName: 'dep-${namePrefix}-evhns-${serviceShort}'
+    location: location
+  }
+}
```
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep:47]:
```suggestion
```

File: avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+          'hidden-title': 'CustomTag'
+        }
+      }
+    ]
+
+    tags: {
+      'hidden-title': 'This is visible in the resource name'
+      Environment: 'Non-Prod'
+      Role: 'DeploymentValidation'
+    }
```
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep:79]:
```suggestion
    ]
```

File: avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    name: '${namePrefix}${serviceShort}001'
+    lock: {
+      kind: 'CanNotDelete'
+      name: 'myCustomLockName'
+    }
+    publicIPPrefixObjects: [
+      {
+        name: '${namePrefix}${serviceShort}001-pippre'
+        prefixLength: 30
+
```
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep:68]:
```suggestion
```

File: avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep (hunk 14 above)
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep:79]:
```suggestion
```

File: avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem' ]: {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
+  params: {
+    location: location
+    name: '${namePrefix}${serviceShort}001'
+    lock: {
+      kind: 'CanNotDelete'
+      name: 'myCustomLockName'
+    }
```
Comment (AlexanderSehr) [inline avm/res/network/nat-gateway/tests/e2e/prefixCombined/main.test.bicep:49]:
```suggestion
```

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]
After the latest changes please re-run the pipeline again and we can merge the PR in :)

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**